api/cloud.proxy.js

```diff
@@ -148,13 +148,44 @@ IoTKitCloud.prototype.update = function(callback) {
         doc.deviceToken = me.secret.deviceToken;
         doc.deviceId = me.deviceId;
         me.logger.info("Updating metadata...");
-        me.proxy.attributes(doc, function () {
-            me.logger.debug("attributes has returned from ", me.proxy.type);
-            if (callback) {
-                callback();
+
+        //creating new variable is necessary cause it can be modified by getDevice function
+        //we need unmodified 'doc' var for update call
+        var docGD= {};
+        if(me.deviceName){
+            docGD.name = me.deviceName;
+        }
+        docGD.deviceToken = me.secret.deviceToken;
+        docGD.deviceId = me.deviceId;
+
+
+        //get device to read existing attributes
+        me.proxy.getDevice(docGD, function(result){
+
+            //delete attributes which agent should update with its own values
+            for(var existingAttributeName in doc.attributes){
+                delete result.attributes[existingAttributeName];
             }
-        });
-        me.logger.info("Metadata updated.");
+
+            //append custom attributes to update body
+            for (var attributeName in result.attributes){
+                doc.attributes[attributeName] = result.attributes[attributeName];
```

Comment:
Why not just make simple if doc.attributes[attributeName]===undefined instead of having two loops?